python/pyspark/sql/connect/session.py

```diff
@@ -385,6 +402,138 @@ def udf(self) -> Any:
+          with JVM via ``spark.plugins`` feature.
+        """
+        session = PySparkSession._instantiatedSession
+        if session is None or session._sc._jsc is None:
+            conf = SparkConf()
+            # Check if we're using unreleased version that is in development.
+            # Also checks SPARK_TESTING for RC versions.
+            is_dev_mode = (
+                "dev" in LooseVersion(__version__).version or "SPARK_TESTING" in os.environ
+            )
+            connect_jar = None
+
+            if is_dev_mode:
+                from pyspark.testing.utils import search_jar
+
+                # Note that, in production, spark.jars.packages configuration should be
+                # set by users. Here we're automatically searching the jars locally built.
+                connect_jar = search_jar(
+                    "connector/connect/server", "spark-connect-assembly-", "spark-connect"
+                )
+                if connect_jar is not None:
+                    origin_jars = conf.get("spark.jars")
+                    if origin_jars is not None:
+                        conf.set("spark.jars", f"{origin_jars},{connect_jar}")
+                    else:
+                        conf.set("spark.jars", connect_jar)
+                else:
+                    warnings.warn(
+                        "Attempted to automatically find the Spark Connect jars because "
+                        "'SPARK_TESTING' environment variable is set, or the current PySpark "
+                        f"version is dev version ({__version__}). However, the jar was not found. "
+                        "Manually locate the jars and specify them, e.g., 'spark.jars' "
+                        "configuration."
+                    )
+
+            conf.set("spark.master", master)
+
+            connect_plugin = "org.apache.spark.sql.connect.SparkConnectPlugin"
+            origin_plugin = conf.get("spark.plugins")
+            if origin_plugin is not None:
+                conf.set("spark.plugins", f"{origin_plugin},{connect_plugin}")
+            else:
+                conf.set("spark.plugins", connect_plugin)
+
+            origin_remote = os.environ.get("SPARK_REMOTE", None)
+            origin_args = os.environ.get("PYSPARK_SUBMIT_ARGS", None)
+            try:
+                if origin_remote is not None:
+                    # So SparkSubmit thinks no remote is set in order to
+                    # start the regular PySpark session.
+                    del os.environ["SPARK_REMOTE"]
+
+                # PySpark shell launches Py4J server from Python.
+                # Remove "--remote" option specified, and use plain arguments.
+                # NOTE that this is not used in regular PySpark application
+                # submission because JVM at this point is already running.
+                os.environ["PYSPARK_SUBMIT_ARGS"] = '"--name" "PySparkShell" "pyspark-shell"'
```

Comment:
I think I should take another look to make sure on passing all extra arguments go through here (although Spark Connect doesn't support all of them ...). Let me revisit this after merging this PR.